conan/tools/microsoft/visual.py

```diff
@@ -100,8 +101,10 @@ def generate(self, scope="build"):
         """
         check_duplicated_generator(self, self._conanfile)
         conanfile = self._conanfile
-        os_ = conanfile.settings.get_safe("os")
-        if os_ != "Windows":
+
+        # Derive subsystem name from the current scope.
+        subsystem = deduce_subsystem(conanfile, scope)
```

Comment:
I have recovered some tests that were pending to migrate after 1.X -> 2.0 migration, and that test will reveal that this is breaking, actually breaking a lot of users that build with Windows subsystems and ``msvc`` as compiler.
I think we need to keep the ``conanfile.settings.get_safe("os") == "Windows"``, and in any case we could consider adding the ``conanfile.settings_build.get_safe("os")==Windows`` as an extra condition to avoid the Linux->Windows cross-build issue

Reply:
The problem is that ``deduce_subsystem()`` returning ``msys2`` and still compiling with ``msvc`` and using ``vcvars.bat`` is a valid scenario. ``VCVars().generate()`` should create a ``conanvcvars.bat`` for this case too. Other generators are using the ``deduce_subsystem()`` mostly to adapt paths, but not to decide if the generator is used or not.